### PR TITLE
Process manager

### PR DESCRIPTION
**Pull request by alpha-beta-soup**
This is not yet ready for merging: apart from the WPS JSON binding spec still being in flux, I'm just making this PR prematurely so there's space for some discussion, review, criticism and hopefully some commits from interested people.

### CODE REVIEW

**Comment by alpha-beta-soup** [comment]
Sample of the HTML templating for a job status:
Screenshot from 2020-03-11 15-31-39

File: pygeoapi/api.py
Comment (alpha-beta-soup) [inline pygeoapi/api.py:2387]: 204 No Content with a body?

File: pygeoapi/util.py
Comment (alpha-beta-soup) [inline pygeoapi/util.py:416]: Is this in spec?

Comment (tomkralidis) [reply pygeoapi/util.py:416]: Looks like it: https://github.com/opengeospatial/ogcapi-processes/blob/954213317742fc0ae769f8c204b3712a047c8099/core/openapi/schemas/statusInfo.yaml#L15

**Comment by tomkralidis** [comment]
Update: thanks much to @alpha-beta-soup, @jorgejesus, and @totycro for their significant contributions to this work.

At long last, I've added some updates, at which point this PR is ready for review:  Notes:
- thinned out advanced HTML functionality (forms, file uploads).  Users looking for advanced HTML are encouraged to implement custom [templates](https://docs.pygeoapi.io/en/latest/html-templating.html)
- created a dummy process manager for default/synchronous execution
- small updates to OpenAPI job and result workflow
- updated tests
- added documentation
- Travis is failing based on Docker errors which #587 looks like it addresses.  Local testing passes
- added Starlette routes

Note that I have NOT rebased to current master (currently shows 79 commits ahead, 12 commits behind master).

TODO:
- implement native file uploads; we need to clarify/get updated with OGC API - Processes how this supported
- raw result file downloads

File: pygeoapi/process/manager/base.py
Comment (totycro) [inline pygeoapi/process/manager/base.py:58]: `create()` and `delete()` are currently unused. `load_plugin()` will just call the constructor.

File: pygeoapi/process/manager/base.py
Comment (totycro) [inline pygeoapi/process/manager/base.py:67]: This is currently only used in the tinydb manager. Maybe leave it out of the base class?

File: pygeoapi/process/manager/base.py
Comment (totycro) [inline pygeoapi/process/manager/base.py:122]: This is currently unused. In api.py, `get_process_job_status()` actually calls `manager.get_job_result()`, so we should probably unify whether we call this `status` or `result`

**Review by totycro — APPROVED** [review]
I like this current structure and I'm in favor of merging this very soon since it constitutes a useful feature sets.
Further extensions/refinements can then still be made, but those shouldn't be breaking changes to what's implemented here.

There's still the obstacle of getting this to run in docker and I'm not completely sure that the other PR (https://github.com/geopython/pygeoapi/pull/587) completely fixes it. The base image will now install Python 3.9 which causes some incompatiblities (I haven't tested this well enough to be able to say anything concrete about it, I just wanted to mention that).

File: pygeoapi/flask_app.py
Comment (alpha-beta-soup) [reply pygeoapi/flask_app.py:429]: `/results` is not `/results/{resource}`

Comment (tomkralidis) [reply pygeoapi/flask_app.py:429]: echo @alpha-beta-soup

File: pygeoapi/process/base.py
Comment (tomkralidis) [reply pygeoapi/process/base.py:73]: Fixed.

File: pygeoapi/process/manager/base.py
Comment (tomkralidis) [reply pygeoapi/process/manager/base.py:58]: I think you mean `create()` and `destroy()`?  Removed/fixed.

File: pygeoapi/process/manager/base.py
Comment (tomkralidis) [reply pygeoapi/process/manager/base.py:67]: Updated.

File: pygeoapi/process/manager/base.py
Comment (tomkralidis) [reply pygeoapi/process/manager/base.py:122]: Fixed.

**Comment by tomkralidis** [comment]
FYI fixes pushed for review/comment.

**Review by tomkralidis — APPROVED** [review]

**Comment by tomkralidis** [comment]
Hi all: I think we're almost there pending some final review.

Having said this, this branch currently has 82 commits over 18 months, and affects 40 files, which includes numerous rebases.  Is there anything to be aware of before we merge so nothing unexpected happens?

File: pygeoapi/util.py
Comment (alpha-beta-soup) [inline pygeoapi/util.py:60]: Unlcear where the boundary is between a manager and the core API? (I think this would be necessary for file upload, but that's absent now.)

File: pygeoapi/util.py
Comment (alpha-beta-soup) [inline pygeoapi/util.py:188]: Remove

File: pygeoapi/util.py
Comment (alpha-beta-soup) [inline pygeoapi/util.py:419]: Could be removed if use of `"finished"` is purged

Comment (alpha-beta-soup) [reply pygeoapi/util.py:419]: which may already be the case, it's possible this comment is out of date now

File: requirements-provider.txt
Comment (alpha-beta-soup) [inline requirements-provider.txt:9]: requirements-processes.txt ?

File: pygeoapi/util.py
Comment (alpha-beta-soup) [reply pygeoapi/util.py:188]: Just line 188, not sure why the rest is highlighted

File: pygeoapi/util.py
Comment (alpha-beta-soup) [reply pygeoapi/util.py:419]: Referring only to `finished`, not `dismissed`

File: pygeoapi/util.py
Comment (tomkralidis) [reply pygeoapi/util.py:188]: Fixed

File: pygeoapi/util.py
Comment (tomkralidis) [reply pygeoapi/util.py:419]: Updated to align to https://github.com/opengeospatial/ogcapi-processes/blob/master/core/openapi/schemas/statusInfo.yaml#L8

File: requirements-provider.txt
Comment (tomkralidis) [reply requirements-provider.txt:9]: Good point.  Once OARec is implemented, TinyDB will also be the backend. At that point we can decide whether to move it into core requirements or not.

File: pygeoapi/util.py
Comment (tomkralidis) [reply pygeoapi/util.py:60]: The core API includes a "dummy" manager (which always forces synchronous execution).

File: pygeoapi/util.py
Comment (tomkralidis) [reply pygeoapi/util.py:59]: Updated.

File: pygeoapi/flask_app.py
Comment (tomkralidis) [reply pygeoapi/flask_app.py:385]: Fixed.

File: pygeoapi-config.yml
Comment (tomkralidis) [reply pygeoapi-config.yml:47]: Just thinking of users with no processes defined not to have to carry a manager along (the dummy manager is passthrough, basically).

File: pygeoapi/templates/process.html
Comment (tomkralidis) [reply pygeoapi/templates/process.html]: Added back.

File: pygeoapi/api.py
Comment (tomkralidis) [reply pygeoapi/api.py]: Simplified logic.

File: pygeoapi/api.py
Comment (tomkralidis) [reply pygeoapi/api.py]: Simplfied logic.

**Pull request merged**
